### PR TITLE
fix(conversations): unarchive and maintain local archives

### PR DESCRIPTION
**Pull request by tlebon**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

_Briefly describe the issue you have solved or implemented with this pull request. If the PR contains multiple issues, use a bullet list._

### Causes (Optional)

_Briefly describe the causes behind the issues. This could be helpful to understand the adopted solutions behind some nasty bugs or complex issues._

### Solutions

_Briefly describe the solutions you have implemented for the issues explained above._

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: src/script/conversation/ConversationMapper.ts [hunk 1]
```diff
@@ -393,7 +402,8 @@ export class ConversationMapper {
 
         return mergedConversation;
       })
-      .concat(failedConversations);
+      .concat(failedConversations as ConversationDatabaseData[])
```
Comment (aweiss-dev) [inline src/script/conversation/ConversationMapper.ts]: Don't do this please. Use type guards to make sure the data is the right one. Assertions are bad :)

**Review by aweiss-dev — CHANGES_REQUESTED** [review]

File: src/script/conversation/ConversationRepository.ts [hunk 2]
```diff
@@ -1943,7 +1943,13 @@ export class ConversationRepository {
         if (conversationEntity) {
           // Check if conversation was archived
           previouslyArchived = conversationEntity.is_archived();
-
+          if (
+            previouslyArchived &&
+            conversationEntity.status() === ConversationStatus.PAST_MEMBER &&
+            type === CONVERSATION_EVENT.MEMBER_JOIN
+          ) {
+            this.unarchiveConversation(conversationEntity, false, 'event from notification stream');
```
Comment (arjita-mitra) [inline src/script/conversation/ConversationRepository.ts]: 'event from notification stream'
Can we make it a constant?

File: src/script/conversation/ConversationMapper.ts [hunk 3]
```diff
@@ -393,7 +403,8 @@ export class ConversationMapper {
 
         return mergedConversation;
       })
-      .concat(failedConversations);
+      .concat(failedConversations)
```
Comment (arjita-mitra) [inline src/script/conversation/ConversationMapper.ts]: Any reason for not using spread here?

Comment (tlebon) [reply src/script/conversation/ConversationMapper.ts]: no, just because ts was annoying me.

Comment (przemvs) [reply src/script/conversation/ConversationMapper.ts]: I'll go with spread by moving `return remoteConversations.found` => `const foundRemoteConversations = ...`
And then `return [...foundRemoteConversations, ...failedConversations, ...localArchives]` :)

File: src/script/conversation/ConversationRepository.ts [hunk 4]
```diff
@@ -1943,7 +1943,13 @@ export class ConversationRepository {
         if (conversationEntity) {
           // Check if conversation was archived
           previouslyArchived = conversationEntity.is_archived();
-
+          if (
+            previouslyArchived &&
+            conversationEntity.status() === ConversationStatus.PAST_MEMBER &&
```
Comment (przemvs) [inline src/script/conversation/ConversationRepository.ts]: I'll move conversationEntity.status() to variable, and all comparision move to variable to make if () more readable.

File: src/script/conversation/ConversationRepository.ts [hunk 5]
```diff
@@ -1943,7 +1943,13 @@ export class ConversationRepository {
...
+            type === CONVERSATION_EVENT.MEMBER_JOIN
```
Comment (przemvs) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
          const isPastMemberStatus = conversationEntity.status() === ConversationStatus.PAST_MEMBER;
          const isMemberJoinType = type === CONVERSATION_EVENT.MEMBER_JOIN;
          
          if (previouslyArchived && isPastMemberStatus && isMemberJoinType
```

Comment (przemvs) [reply src/script/conversation/ConversationRepository.ts]: Smth like that :)

File: src/script/conversation/ConversationRepository.ts (hunk 2 above)
Comment (tlebon) [reply src/script/conversation/ConversationRepository.ts]: i added it to the class as a static member

**Review by przemvs — APPROVED** [review]

File: src/script/conversation/ConversationMapper.ts (hunk 1 above)
Comment (tlebon) [reply src/script/conversation/ConversationMapper.ts]: gone

**Review by aweiss-dev — APPROVED** [review]

**Pull request merged**